compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirDeclarationsResolveTransformer.kt

```diff
@@ -510,6 +517,47 @@ open class FirDeclarationsResolveTransformer(transformer: FirBodyResolveTransfor
         return result.compose()
     }
 
+    /*
+     * Suppose a function without an explicit return type just returns an anonymous object:
+     *
+     *   fun foo(...) = object : ObjectSuperType {
+     *     override fun ...
+     *   }
+     *
+     * Without unwrapping, the return type ended up with that anonymous object (<no name provided>), while the resolved super type, which
+     * acts like an implementing interface, is a better fit. In fact, exposing an anonymous object types is prohibited for certain cases,
+     * e.g., KT-33917. We can also apply this to any local types.
+     */
+    private fun FirTypeRef.hideLocalTypeIfNeeded(
+        containingCallableVisibility: Visibility?,
+        isInlineFunction: Boolean = false
+    ): FirTypeRef {
+        if (containingCallableVisibility == null) {
+            return this
+        }
+        // Approximate types for non-private (all but package private or private) members.
+        // Also private inline functions, as per KT-33917.
+        if (containingCallableVisibility == Visibilities.Public ||
+            containingCallableVisibility == Visibilities.Protected ||
+            containingCallableVisibility == Visibilities.Internal ||
```

Comment:
Somehow an `internal` _property_'s local type is not hidden. See unambiguousObjectExpressionType.fir.kt#L30 and #L140. (This is unresolved in old frontend.)

Reply:
Good catch! It turned out that I missed some other places where property's return type is updated. Correctly unresolved now. Thanks!